business/namespaces.go

```diff
@@ -78,6 +80,26 @@ func (in *NamespaceService) GetNamespaces(ctx context.Context) ([]models.Namespa
 
 	configObject := config.Get()
 
+	var istioConfig *kubernetes.IstioMeshConfig
+	if kialiCache != nil {
+		// we will need the Istio discoverySelectors
+		icm, err := kialiCache.GetConfigMap(configObject.IstioNamespace, configObject.ExternalServices.Istio.ConfigMapName)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to get the Istio ConfigMap: %v", err)
+		}
+		ic, err := kubernetes.GetIstioConfigMap(icm)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to parse the Istio ConfigMap: %v", err)
+		}
+		istioConfig = ic
+	}
+
+	if istioConfig.DiscoverySelectors != nil && len(istioConfig.DiscoverySelectors) > 0 {
```

Comment:
Just trying to get your tests passing... you likely need an additional check for `istioConfig != nil`

Reply:
Well, this is what I'm struggling with. In production, the Kiali server should always have access to the istio configmap - that value should never be nil (or, technically, you should never get here if its nil). This function aborts with an error if the configmap cannot be processed - so you should never get here.

Obviously, I could just not abort and keep going with a nil ConfigMap but:

(a) that would solely be to support unit testing and is not valid in production - and that sucks to have production code written in a way solely to support unit tests, especially when it really isn't "right"

(b) IMO, if we can't get the DS, then we need to abort. This could be considered a security issue. Because suppose Kiali is configured with ** but the user wants to only make available those namespaces filtered by DS - if we do not have access to what the DSs are - we should NOT default to all namespace (i.e. we should not default to a list of namespaces that were not filtered by DS). To me, that is a security flaw. If we can't get to the DS, then we need to abort and tell the user we are broken until Kiali has access to see the Istio ConfigMap so it can correctly display only those namespaces we should be.

Reply:
> Well, this is what I'm struggling with. In production, the Kiali server should always have access to the istio configmap - that value should never be nil (or, technically, you should never get here if its nil). This function aborts with an error if the configmap cannot be processed - so you should never get here.

Kiali may never have access to external control planes or istio may temporarily be down in which case Kiali should still keep functioning.

> This could be considered a security issue.

I thought DS was not about access? Isn't that what `cluster_wide_access` is for? DS is not a security boundary for istio either I don't think.

Reply:
If we all don't think this is a security-issue to worry about, I can relax that code. I'm security-paranoid because I know all the times people have complained to us that we expose more than we should, which is why I was hesitant.

Reply:
This has been reworked so the server continues on even if the Istio ConfigMap is not found or for some reason can't be parsed. It will still log an error but it keeps going.

Reply:
lo-and-behold almost all the unit tests pass again :)

There is one failure that I don't understand:
```
--- FAIL: TestTlsPerfNsDr (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x18022ae]

goroutine 8785 [running]:
testing.tRunner.func1.2({0x1c0ef60, 0x311c200})
	/home/jmazzite/bin/go-install/go/src/testing/testing.go:1396 +0x24e
testing.tRunner.func1()
	/home/jmazzite/bin/go-install/go/src/testing/testing.go:1399 +0x39f
panic({0x1c0ef60, 0x311c200})
	/home/jmazzite/bin/go-install/go/src/runtime/panic.go:884 +0x212
github.com/kiali/kiali/kubernetes/cache.(*kialiCacheImpl).GetConfigMap(0xc000818790?, {0x1ef387a?, 0x1ef1e16?}, {0x1ee9965?, 0xc000b42a00?})
	<autogenerated>:1 +0x2e
github.com/kiali/kiali/business.(*NamespaceService).GetNamespaces(0xc000a30e70, {0x21fa8f0, 0xc0001ac008})
	/home/jmazzite/source/kiali/kiali/business/namespaces.go:86 +0x278
github.com/kiali/kiali/business.(*TLSService).getNamespaces(0x21fa8f0?, {0x21fa8f0?, 0xc0001ac008?})
```


I don't get it because this is wrapped in the kialiCache != nil test, so something else must be blowing up

Reply:
this test has been fixed with a not-so-obvious change.